dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -3510,13 +3541,77 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(
 {
     sanitizeCheckReadRanges(__FUNCTION__, read_ranges, rowkey_range, log);
 
-    auto bitmap_filter = buildBitmapFilter(
+    BitmapFilterPtr bitmap_filter = nullptr;
+    if (executor && executor->column_range && executor->column_range->type != ColumnRangeType::Unsupported)
+    {
+        bool all_dmfile_packs_skipped
+            = std::all_of(pack_filter_results.begin(), pack_filter_results.end(), [](const auto & res) {
+                  return res->countUsePack() == 0;
+              });
+        if (!all_dmfile_packs_skipped)
+        {
+            bitmap_filter = InvertedIndexReaderFromSegment::loadStable(
+                segment_snap,
+                executor->column_range,
+                dm_context.global_context.getLightLocalIndexCache(),
+                dm_context.scan_context);
+            size_t skipped_pack = modifyPackFilterResults(segment_snap, pack_filter_results, bitmap_filter);
+            dm_context.scan_context->inverted_idx_search_skipped_packs += skipped_pack;
+            LOG_DEBUG(
+                segment_snap->log,
+                "Finish load inverted index, column_range={}, bitmap_filter={}/{}, skipped_pack={}",
+                executor->column_range->toDebugString(),
+                bitmap_filter->count(),
+                bitmap_filter->size(),
+                skipped_pack);
+        }
+        else
+        {
+            LOG_DEBUG(
+                segment_snap->log,
+                "Skip load inverted index, all dmfile packs are skipped, column_range={}",
+                executor->column_range->toDebugString());
+        }
+    }
+
+    auto mvcc_bitmap_filter = buildBitmapFilter(
         dm_context,
         segment_snap,
         read_ranges,
         pack_filter_results,
         start_ts,
         build_bitmap_filter_block_rows);
+    if (bitmap_filter)
+    {
+        if (!readStableOnly(dm_context, segment_snap))
+        {
+            auto delta_index_bitmap = InvertedIndexReaderFromSegment::loadDelta(
+                segment_snap,
+                executor->column_range,
+                dm_context.global_context.getLightLocalIndexCache(),
+                dm_context.scan_context);
+            bitmap_filter->append(*delta_index_bitmap);
+        }
+
```

Comment:
Should these lines put right after `InvertedIndexReaderFromSegment::loadStable`?

Reply:
Currently, we do not support skip delta files, so putting them right after `InvertedIndexReaderFromSegment::loadStable` is ok.

But if we do support skip delta files, loading delta bitmap before MVCC may cause correctness issues.

So I think putting here is better.

Reply:
OK, I get what you mean. Only after bitAnd `bitmap_filter.logicalAnd(mvcc_bitmap)`, we can ensure the correctness of skipping delta files.